File: Editor/MapEditor/object3dcustom.cpp
/*
    RPG Paper Maker Copyright (C) 2017-2021 Wano

    RPG Paper Maker engine is under proprietary license.
    This source code is also copyrighted.

    Use Commercial edition for commercial use of your games.
    See RPG Paper Maker EULA here:
        http://rpg-paper-maker.com/index.php/eula.
*/

#include <QtMath>
#include "object3dcustom.h"
#include "rpm.h"
#include "common.h"

// -------------------------------------------------------
//
//  CONSTRUCTOR / DESTRUCTOR / GET / SET
//
// -------------------------------------------------------

Object3DCustomDatas::Object3DCustomDatas() :
    Object3DDatas()
{

}

Object3DCustomDatas::Object3DCustomDatas(SystemObject3D *datas) :
    Object3DDatas(datas)
{
    m_min = datas->obj()->minVertex();
    m_max = datas->obj()->maxVertex();
    m_box = QBox3D(m_min, m_max);
}

Object3DCustomDatas::~Object3DCustomDatas()
{

}

// -------------------------------------------------------
//
//  INTERMEDIARY FUNCTIONS
//
// -------------------------------------------------------

void Object3DCustomDatas::rotateVertex(QVector3D& vec, Position &position,
    QVector3D& center)
{
    if (position.angleY() != 0.0) {
        SpriteDatas::rotateVertexX(vec, center, position.angleY(), 0, 1, 0);
    }
    if (position.angleX() != 0.0) {
        SpriteDatas::rotateVertexX(vec, center, position.angleX(), 1, 0, 0);
    }
    if (position.angleZ() != 0.0) {
        SpriteDatas::rotateVertexX(vec, center, position.angleZ(), 0, 0, 1);
    }
}

// -------------------------------------------------------
//
//  VIRTUAL FUNCTIONS
//
// -------------------------------------------------------

qreal Object3DCustomDatas::xMin() const {
    return static_cast<qreal>(m_min.x()) / RPM::get()->getSquareSize();
}

// -------------------------------------------------------

qreal Object3DCustomDatas::yMin() const {
    return static_cast<qreal>(m_min.y()) / RPM::get()->getSquareSize();
}

// -------------------------------------------------------

qreal Object3DCustomDatas::zMin() const {
    return static_cast<qreal>(m_min.z()) / RPM::get()->getSquareSize();
}

// -------------------------------------------------------

qreal Object3DCustomDatas::xMax() const {
    return static_cast<qreal>(m_max.x()) / RPM::get()->getSquareSize();
}

// -------------------------------------------------------

qreal Object3DCustomDatas::yMax() const {
    return static_cast<qreal>(m_max.y()) / RPM::get()->getSquareSize();
}

// -------------------------------------------------------

qreal Object3DCustomDatas::zMax() const {
    return static_cast<qreal>(m_max.z()) / RPM::get()->getSquareSize();
}

// -------------------------------------------------------

int Object3DCustomDatas::width() const {
    return Common::ceil(static_cast<qreal>(this->widthPixels()) / RPM::get()
        ->getSquareSize());
}

// -------------------------------------------------------

int Object3DCustomDatas::height() const {
    return Common::ceil(static_cast<qreal>(this->heightPixels()) / RPM::get()
        ->getSquareSize());
}

// -------------------------------------------------------

int Object3DCustomDatas::depth() const {
    return Common::ceil(static_cast<qreal>(this->depthPixels()) / RPM::get()
        ->getSquareSize());
}

// -------------------------------------------------------

int Object3DCustomDatas::widthPixels() const {
    QVector3D minimum, maximum;

    minimum = m_box.minimum();
    maximum = m_box.maximum();

    return static_cast<int>(maximum.x() - minimum.x());
}

// -------------------------------------------------------

int Object3DCustomDatas::heightPixels() const {
    QVector3D minimum, maximum;

    minimum = m_box.minimum();
    maximum = m_box.maximum();

    return static_cast<int>(maximum.y() - minimum.y());
}

// -------------------------------------------------------

int Object3DCustomDatas::depthPixels() const {
    QVector3D minimum, maximum;

    minimum = m_box.minimum();
    maximum = m_box.maximum();

    return static_cast<int>(maximum.z() - minimum.z());
}

// -------------------------------------------------------

int Object3DCustomDatas::xOverflowCenter() const
{
    return qFloor(static_cast<qreal>(m_box.center().x() / RPM::getSquareSize()));
}

// -------------------------------------------------------

int Object3DCustomDatas::yOverflowCenter() const
{
    return qFloor(static_cast<qreal>(m_box.center().y() / RPM::getSquareSize()));
}

// -------------------------------------------------------

int Object3DCustomDatas::zOverflowCenter() const
{
    return qFloor(static_cast<qreal>(m_box.center().z() / RPM::getSquareSize()));
}

// -------------------------------------------------------

void Object3DCustomDatas::getCenter(QVector3D &center) const {
    QVector3D vec;

    vec = m_box.center();
    center.setX(vec.x());
    center.setY(vec.y());
    center.setZ(vec.z());
}

// -------------------------------------------------------

int Object3DCustomDatas::minDistanceFromCenter() const {
    int w, h, d, max;

    w = this->widthPixels();
    h = this->heightPixels();
    d = this->depthPixels();
    max = std::max(w, h);
    max = std::max(max, d);

    return max + ((w + h + d) / 2);
}

// -------------------------------------------------------

void Object3DCustomDatas::initializeVertices(QVector<Vertex> &vertices,
    QVector<GLuint> &indexes, Position &position, unsigned int &count)
{
    QVector3D positionOffset, size, center, vec, vecA, vecB;
    SystemCustomShape *shape;
    QPair<int, int> index;
    int i, l, squareSize;
    float scale;

    scale = static_cast<float>(m_datas->scale());
    squareSize = RPM::get()->getSquareSize();
    size.setX(scale * squareSize);
    size.setY(scale * squareSize);
    size.setZ(scale * squareSize);
    shape = m_datas->obj();
    positionOffset.setX((position.x() * squareSize) + (squareSize / 2));
    positionOffset.setY(position.getY(squareSize));
    positionOffset.setZ((position.z() * squareSize) + (squareSize / 2));
    m_box = QBox3D(shape->minVertex() * size + positionOffset, shape
        ->maxVertex() * size + positionOffset);
    this->getCenter(center);
    for (i = 0, l = shape->facesCount(); i < l; i++) {
        index = shape->getFace(i);
        vec = shape->getVertexAt(index.first) * size + positionOffset;
        this->rotateVertex(vec, position, center);
        vertices.append(Vertex(vec, shape->getTextureAt(index.second)));
        indexes.append(count++);
    }

    // Create collision box according to vertex positions
    vecA = shape->minVertex() * size + positionOffset;
    vecB = shape->maxVertex() * size + positionOffset;
    this->rotateVertex(vecA, position, center);
    this->rotateVertex(vecB, position, center);
    m_box = QBox3D(vecA, vecB);
}

// -------------------------------------------------------

float Object3DCustomDatas::intersection(QRay3D &ray) const {
    float minDistance = 0;

    Common::getMinDistance(minDistance, m_box.intersection(ray));

    return minDistance;
}

// -------------------------------------------------------

void Object3DCustomDatas::read(const QJsonObject & json) {
    Object3DDatas::read(json);
}

// -------------------------------------------------------

void Object3DCustomDatas::write(QJsonObject & json) const{
    Object3DDatas::write(json);
}

File: EditorApp/Content/example/Datas/Scripts/System/Core/Object3DCustom.js
/*
    RPG Paper Maker Copyright (C) 2017-2022 Wano

    RPG Paper Maker engine is under proprietary license.
    This source code is also copyrighted.

    Use Commercial edition for commercial use of your games.
    See RPG Paper Maker EULA here:
        http://rpg-paper-maker.com/index.php/eula.
*/
import { Datas } from "../index.js";
import { Enum } from "../Common/index.js";
var CustomShapeKind = Enum.CustomShapeKind;
var ObjectCollisionKind = Enum.ObjectCollisionKind;
import { Sprite } from "./Sprite.js";
import { Object3D } from "./Object3D.js";
import { Vector3 } from "./Vector3.js";
import { CustomGeometry } from "./CustomGeometry.js";
/** @class
 *  A 3D object custom in the map.
 *  @extends Object3D
 *  @param {Record<string, any>} json - Json object describing the object 3D custom
 *  @param {System.Object3D} datas - The System object 3D
 */
class Object3DCustom extends Object3D {
    constructor(json, datas) {
        super();
        this.datas = datas;
        if (json) {
            this.read(json);
        }
    }
    /**
     *  Create a new 3D object box.
     *  @static
     *  @param {System.Object3D} datas - The object datas
     *  @returns {Core.Object3DBox}
     */
    static create(datas) {
        let object = new Object3DCustom(undefined, datas);
        object.id = datas.id;
        return object;
    }
    /**
     *  Read the JSON associated to the object 3D custom.
     *  @param {Record<string, any>} json - Json object describing the object 3D
     *  custom
     */
    read(json) {
        super.read(json);
        this.id = this.datas.id;
    }
    /**
     *  Get the center vector.
     *  @returns {Vector3}
     */
    getCenterVector() {
        return Datas.Shapes.get(Enum.CustomShapeKind.OBJ, this.datas.objID)
            .geometry.center.clone();
    }
    /**
     *  Update the geometry of a group of objects 3D cutom with the same
     *  material.
     *  @param {Core.CustomGeometry} geometry - Geometry of the object 3D custom
     *  @param {Position} position - The position of the object 3D custom
     *  @param {number} count - The faces count
     *  @return {any[]}
    */
    updateGeometry(geometry, position, count) {
        let localPosition = position.toVector3();
        let modelGeometry = Datas.Shapes.get(CustomShapeKind.OBJ, this
            .datas.objID).geometry;
        let vertices = modelGeometry.vertices;
        let uvs = modelGeometry.uvs;
        let center = modelGeometry.center.clone();
        let scale = this.datas.scale;
        let scaleVec = new Vector3(scale * position.scaleX, scale * position
            .scaleY, scale * position.scaleZ);
        let angleY = position.angleY;
        let angleX = position.angleX;
        let angleZ = position.angleZ;
        let vecA, vecB, vecC;
        for (let i = 0, l = modelGeometry.vertices.length; i < l; i += 3) {
            vecA = vertices[i].clone();
            vecB = vertices[i + 1].clone();
            vecC = vertices[i + 2].clone();
            vecA.multiply(scaleVec);
            vecB.multiply(scaleVec);
            vecC.multiply(scaleVec);
            if (angleY !== 0.0) {
                Sprite.rotateVertex(vecA, center, angleY, Sprite.Y_AXIS);
                Sprite.rotateVertex(vecB, center, angleY, Sprite.Y_AXIS);
                Sprite.rotateVertex(vecC, center, angleY, Sprite.Y_AXIS);
            }
            if (angleX !== 0.0) {
                Sprite.rotateVertex(vecA, center, angleX, Sprite.X_AXIS);
                Sprite.rotateVertex(vecB, center, angleX, Sprite.X_AXIS);
                Sprite.rotateVertex(vecC, center, angleX, Sprite.X_AXIS);
            }
            if (angleZ !== 0.0) {
                Sprite.rotateVertex(vecA, center, angleZ, Sprite.Z_AXIS);
                Sprite.rotateVertex(vecB, center, angleZ, Sprite.Z_AXIS);
                Sprite.rotateVertex(vecC, center, angleZ, Sprite.Z_AXIS);
            }
            vecA.add(localPosition);
            vecB.add(localPosition);
            vecC.add(localPosition);
            geometry.pushTriangleVertices(vecA, vecB, vecC);
            geometry.pushTriangleIndices(count);
            geometry.pushTriangleUVs(uvs[i].clone(), uvs[i + 1].clone(), uvs[i
                + 2].clone());
            count += 3;
        }
        // Collisions
        let objCollision = new Array;
        if (this.datas.collisionKind === ObjectCollisionKind.Simplified) {
            let obj = this.datas.getObj().geometry;
            let w = obj.w * scale * position.scaleX;
            let h = obj.h * scale * position.scaleY;
            let d = obj.d * scale * position.scaleZ;
            let minPos = obj.minVertex.clone();
            minPos.multiply(scaleVec);
            objCollision.push({
                p: position,
                l: localPosition,
                b: [
                    localPosition.x + minPos.x + (w / 2),
                    localPosition.y + minPos.y + (h / 2),
                    localPosition.z + minPos.z + (d / 2),
                    w,
                    h,
                    d,
                    angleY,
                    angleX,
                    angleZ
                ],
                c: center,
                w: Math.ceil(w / 2 / Datas.Systems.SQUARE_SIZE),
                h: Math.ceil(h / 2 / Datas.Systems.SQUARE_SIZE),
                d: Math.ceil(d / 2 / Datas.Systems.SQUARE_SIZE),
                m: Math.max(Math.max(Math.ceil(w / 2 / Datas.Systems.SQUARE_SIZE), Math.ceil(h / 2 / Datas.Systems.SQUARE_SIZE)), Math.ceil(d / 2 / Datas.Systems.SQUARE_SIZE)),
                k: true
            });
        }
        else if (this.datas.collisionKind === ObjectCollisionKind.Custom) {
            let obj = Datas.Shapes.get(CustomShapeKind.Collisions, this.datas
                .collisionCustomID).geometry;
            let w = obj.w * scale * position.scaleX;
            let h = obj.h * scale * position.scaleY;
            let d = obj.d * scale * position.scaleZ;
            let minPos = obj.minVertex.clone();
            minPos.multiply(scaleVec);
            objCollision.push({
                id: this.datas.collisionCustomID,
                p: position,
                l: localPosition,
                b: [
                    localPosition.x + minPos.x + (w / 2),
                    localPosition.y + minPos.y + (h / 2),
                    localPosition.z + minPos.z + (d / 2),
                    w,
                    h,
                    d,
                    angleY,
                    angleX,
                    angleZ
                ],
                c: center,
                w: Math.ceil(w / 2 / Datas.Systems.SQUARE_SIZE),
                h: Math.ceil(h / 2 / Datas.Systems.SQUARE_SIZE),
                d: Math.ceil(d / 2 / Datas.Systems.SQUARE_SIZE),
                m: Math.max(Math.max(Math.ceil(w / 2 / Datas.Systems.SQUARE_SIZE), Math.ceil(h / 2 / Datas.Systems.SQUARE_SIZE)), Math.ceil(d / 2 / Datas.Systems.SQUARE_SIZE)),
                k: true
            });
        }
        return [count, objCollision];
    }
    /**
     *  Create a new geometry.
     *  @param {Position} position - The position of object 3D
     *  @return {[Core.CustomGeometry, [number, StructMapElementCollision[]]]}
     */
    createGeometry(position) {
        let geometry = new CustomGeometry();
        let collisions = this.updateGeometry(geometry, position, 0);
        geometry.updateAttributes();
        return [geometry, collisions];
    }
}
export { Object3DCustom };
